cpp_ipu_dir/codelets.cpp

```diff
@@ -0,0 +1,25 @@
+#include <poplar/Vertex.hpp>
+
+using namespace poplar;
+
+class VertexPi : public Vertex {
+public:
+    Input<unsigned> id;
+    Input<unsigned> n;
+    Input<float> slice;
+    Output<Vector<float>> out;
+    Output<Vector<unsigned>> cycles;
+
+    bool compute() {
+        float sum = 0;
+        unsigned start = __builtin_ipu_get_scount_l();
+        for (unsigned j=id * n; j <= (id + 1) * n - 1; j++) {
+            float x = (j - 0.5f) * slice;
+            sum += 4 / (1 + x * x);
```

Comment:
I'm honestly struggling to see how any of the elements of `sums` can be negative (let alone their sum be negative) as each element is the sum of strictly positive numbers. I probably did something incredibly dumb which I can't spot for the life of me.

Reply:
Alright, I believe the problem was that I was using integer literals instead of single precision floating point literals (`4` and `1` vs `4.0` and `1.0`), which was screwing everything.